sys: Add wakeup source subsystem

bjarki-andreasen opened the pull request:
## Introduction
Wakeup sources are events which can wake up the system from a sleep state, like what is entered using `poweroff()`. A common wakeup source is a button press through a GPIO, or an RTC alarm raising an interrupt. Zephyr currently has no generic way of defining, enabling or disabling wakeup sources, this is what is addressed with this PR.

The PR draws heavy inspiration from Linux, see [wakeup-source.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/power/wakeup-source.txt)

## Glossary
- Wakeup source: A collection of wakeup events tied to a specific devicetree node
- Wakeup event: An event capable of  waking up the SOC from poweroff
- Poweroff: A state in which an SOC is powered down

Note: Wakeup source defined here is not the same as the `wakeup-source` binding in `pm.yaml`.

## Overview
We have the following GPIO key defined in the devicetree snippet:
```
	gpio_keys {
		compatible = "gpio-keys";
		user_button: button {
			label = "User";
			gpios = <&gpio0 10 (GPIO_ACTIVE_HIGH)>;
			zephyr,code = <INPUT_KEY_0>;
		};
	};
```
We want this GPIO key to be able to wake up the system, so we add the `wakeup-source` property to it to define it as wakeup source:
```
	gpio_keys {
		compatible = "gpio-keys";
		user_button: button {
			label = "User";
			gpios = <&gpio0 10 (GPIO_ACTIVE_HIGH)>;
			zephyr,code = <INPUT_KEY_0>;
			wakeup-source;
		};
	};
```
From the application, we want to get a handle to the wakeup source:
```
static const struct *wakeup_source user_button_ws = WAKEUP_SOURCE_DT_GET(DT_NODELABEL(user_button));
```
Lastly, we want to enable it before powering down:
```
	sys_wakeup_source_enable(user_button_ws);
	sys_poweroff();
```
Enabling the wakeup source using `wakeup_source_enable()` sets a flag indicating that is should be enabled before power off. Once `poweroff()` is called, `z_sys_wakeup_sources_configure()` is called automatically, configuring all enabled power sources.

## How does it work?
For each node which is defined as a wakeup source using the `wakeup-source` property, the following information is stored:
* Every IRQ number
* Every GPIO defined by the property `interrupt-gpios` and the `gpios` property of a `"gpio-keys"` button
* Every vendor specific wakeup source identifier, if any

These properties are stored in the following structure, which is created for each wakeup source
```
struct wakeup_source {
	/** Vendor specific wakeup source identifiers */
	uint16_t *wakeup_ids;
	/** Number of vendor specific wakeup source identifier */
	uint16_t wakeup_ids_size;
	/** Wakeup interrupt numbers */
	uint16_t *wakeup_irq_numbers;
	/** Number of wakeup interrupt numbers */
	uint16_t wakeup_irq_numbers_size;
	/** Wakeup GPIOs */
	struct gpio_dt_spec *wakeup_gpios;
	/** Number of wakeup GPIOs */
	uint16_t wakeup_gpios_size;
	/** Name of wakeup source */
	const char *name;
};
```
The wakeup subsystem will use the stored information, and track which wakeup sources the application has enabled, to then pass this information to the SOC through hooks which may be implemented by the SOC.

### Wakeup ids
The vendor may define a set of SOC specific identifiers which will be passed by the wakeup source subsystem if the wakeup source shall be enabled. The SOC may implement the hook `int z_sys_wakeup_source_enable_id(uint16_t id)` to enable this functionality,

### Wakeup GPIOs
GPIOs used by enabled wakeup sources. The SOC may implement the hook `int z_sys_wakeup_source_enable_gpio(const struct gpio_dt_spec *gpio)` to enable this functionality,

### Wakeup IRQ numbers
IRQ numbers used by enabled wakeup sources. The SOC may implement the hook `int z_sys_wakeup_source_enable_irq(uint16_t irq_number)` to enable this functionality,

## Examples of wakeup sources from devicetree nodes
### GPIO keys
```
	gpio_keys {
		compatible = "gpio-keys";
		user_button: button {
			label = "User";
			gpios = <&gpio0 10 (GPIO_ACTIVE_HIGH)>;
			zephyr,code = <INPUT_KEY_0>;
		};
	};
```
becomes:
```
const struct gpio_dt_spec __wakeup_source_0_wakeup_gpios[] = {
	{.port = __device_dts_ord, .pin = 10, dt_flags = 1},
};

const struct wakeup_source __wakeup_source_0 = {
	.wakeup_gpios = __wakeup_source_1_wakeup_gpios, 
	.wakeup_gpios_size = 1,
};
```
### Sensor with single GPIO routed interrupt
```
	&spi0 {
		sensor: bmi323@0 {
			compatible = "bosch,bmi323"
			reg = <0>;
			interrupt-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>;
			reset-gpios = <&gpio0 12 (GPIO_ACTIVE_LOW)>;
		};
	};
```
becomes:
```
const struct gpio_dt_spec __wakeup_source_1_wakeup_gpios[] = {
	{.port = __device_dts_ord, .pin = 11, dt_flags = 1},
};

const struct wakeup_source __wakeup_source_1 = {
	.wakeup_gpios = __wakeup_source_1_wakeup_gpios,
	.wakeup_gpios_size = 1,
};
```
### Sensor with multiple GPIO routed interrupt with specified events routed to each interrupt (using names)
```
	&spi0 {
		sensor: bmi323@0 {
			compatible = "bosch,bmi323"
			reg = <0>;
			interrupt-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>,
                                          <&gpio1 8 (GPIO_ACTIVE_HIGH)>;
			interrupt-gpio-names = "data_ready", "any_motion";
			reset-gpios = <&gpio0 12 (GPIO_ACTIVE_LOW)>;
		};
	};
```
becomes:
```
const struct gpio_dt_spec __wakeup_source_1_wakeup_gpios[] = {
	{.port = __device_dts_ord, .pin = 11, dt_flags = 1},
	{.port = __device_dts_ord, .pin = 8, dt_flags = 1},
};

const struct wakeup_source __wakeup_source_1 = {
	.wakeup_gpios = __wakeup_source_1_wakeup_gpios,
	.wakeup_gpios_size = 2,
};
```
### Vendor specific RTC peripheral
.dtsi
```
rtc: rtc@400e1860 {
	compatible = "atmel,sam-rtc";
	reg = <0x400e1860 0x100>;
	interrupts = <2 0>;
	wakeup-source-ids = <5>;
	alarms-count = <1>;
	status = "disabled";
};
```
.overlay
```
&rtc {
	status = "okay";
	wakeup-source;
};
```
becomes:
```
uint16_t __wakeup_source_1_wakeup_ids[] = {
	5
};

uint16_t __wakeup_source_1_wakeup_irq_numbers[] = {
	2
};

const struct wakeup_source __wakeup_source_1 = {
	.wakeup_ids = __wakeup_source_1_wakeup_ids, 
	.wakeup_ids_size = 1,
	.wakeup_irq_numbers = __wakeup_source_1_wakeup_irq_numbers, 
	.wakeup_irq_size = 1,
};
```

bjarki-andreasen committed:
sys: Add wakeup source subsystem

Wakeup sources can be identified by the wakeup-source
devicetree property, the properties required to enable and
disable a wakeup-source on a soc level can then be
extracted from the identified wakeup source devicetree
nodes.

This commit adds a structure to store the information
required to enable a wakeup-source, and the required utils
to define these structures.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
@fabiobaltieri @gmarull @nandojve This is a proposal for a generic wakeup source subsystem which matches the Linux implementation quite well :)

bjarki-andreasen committed:
soc: nordic: add naive GPIO wakeup source support

Very simple working implementation of setting up a GPIO as a wakeup
source.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
dts: bindings: input: Add wakeup-source prop to gpio-keys

Add the wakeup-source property to the gpio-keys.yaml
bindings file. This allows for the key to be detected and
used by the wakeup source subsystem.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
sys: poweroff: Add wakeup source configuration

Configure all enabled wakeup sources right before powering
down as part of the poweroff() API.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

fabiobaltieri commented:
Cool! Thanks for working on a generic approach for this, added few people I think worked or are working on wakeup functionalities.

I like that the approach covers gpio and internal devices as well, the case with multiple potential gpios is unclear to me, I see your comment about `gpio-names` but I'm convinced about that... maybe it should just be left up to the driver? I imagine like one of those accelerometer with two interrupt pins, it could have a specific property to pick which one if not both to enable.

bjarki-trackunit commented:
@fabiobaltieri I definitely agree it is up to the driver to define which GPIOs if any to configure if used as a wakeup source. I think we could use a "reserved" gpios name, like `int-gpios` (matches many sensors already)
```
	int-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>;
```
If there are multiple interrupt (wakeup source) GPIOs
```
	int-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>,
		    <&gpio1 8 (GPIO_ACTIVE_HIGH)>;
```
We could even have a special property along the lines of `wakeup-source-gpios` in parallel with the existing gpios (I don't think its a good idea as the user needs to add this in the dts overlay)
```
	foo-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>,
	bar-gpios = <&gpio1 6 (GPIO_ACTIVE_HIGH)>,
	/* Clone of foo-gpios */
	wakeup-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>;
```

erwango commented on dts/bindings/base/pm.yaml:
```diff
@@ -16,6 +16,11 @@ properties:
       Wake up capable devices are disabled (interruptions will not wake up
       the system) by default but they can be enabled at runtime if necessary.
 
+  wakeup-source-ids:
+    type: array
+    description: |
+      Property to identify vendor specific wakeup sources.
```
On STM32, there is a complete decorrelation between wakeup pin numbering and GPIO. This is an information that you need to get from the datasheet. For instance:
![wkup_2](https://github.com/zephyrproject-rtos/zephyr/assets/18574308/e3e400e8-aa66-43fb-b987-db67cb7d541c)
Then control is done using the wakeup pin numbering through dedicated registers (non related to GPIO)
![wkup_pwr_reg](https://github.com/zephyrproject-rtos/zephyr/assets/18574308/a8a36fb3-3dbf-44fb-8c9a-3930bb56a0ad)
This is typically the thing useful to get from device tree rather than hardcoding a table to be updated for each and other STM32 variant.

Can this property be used to convey this piece of information ?

bjarki-trackunit replied:
Yep :) pretty much exactly what it's for

bjarki-trackunit replied:
Not really, device pins are used in the dts overlays like this
```
sensor: bmi323@0 {
	compatible = "bosch,bmi323"
	reg = <0>;
	gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>;
	wakeup-source;
}
```
so the user would have to look up the mapping and apply it manually, it is intended for peripherals defined in the .dtsi, like an RTC for example.
```
rtc: rtc@4000000 {
	compatible = "atmel,sam-rtc";
	reg = <4000000 0x100>;
	interrupts = <2 0>,
	             <3 0>;
	interrupt-names = "update", "wakeup";
	alarms-count = <1>;
	status = "disabled";
};
```

Not really, the GPIOs are defined like this

erwango replied:
> Not really, device pins are used in the dts overlays like this

Ok, got it.

> The SOC will have to implement int z_sys_wakeup_source_enable_gpio(const struct gpio_dt_spec *gpio) and keep some internal header or similar to map the port + pin to the wakeup number

Then, if we can't use `wakeup-source-ids`, we should be able to find a way to use device tree to describe this mapping, right ?

henrikbrixandersen commented:
> think we could use a "reserved" gpios name, like `int-gpios` or `irq-gpios` (matches some sensors already)

I don't believe that would be a good design choice. It would potentially limit compatibility with Linux devicetree bindings and "magically" imply functionality that may or may not be desired for the given binding.

bjarki-trackunit commented:
> I don't believe that would be a good design choice. It would potentially limit compatibility with Linux devicetree bindings and "magically" imply functionality that may or may not be desired for the given binding.

Do you happen to have some examples from Linux regarding wakeup from a gpio outside of the "gpio-keys" compatible, I have not found any yet

nashif commented:
@bjarki-trackunit I do not think this should be submitted as a new subsystem, please fold this under subsys/pm somehow, this is all about power management after all, it does not need to depend on current PM infrastrucutre.

bjarki-trackunit commented:
> @bjarki-trackunit I do not think this should be submitted as a new subsystem, please fold this under subsys/pm somehow, this is all about power management after all, it does not need to depend on current PM infrastrucutre.

The wakeup source subsystem does not manage devices like the pm subsystem, it configures the SOC to wake up on events, similarly `sys_poweroff()` configures the SOC to go to sleep. It is an extension of the SOC capabilities, not an addition to device drivers.

Like  `sys_poweroff()`, it does not depend on, or is even related to PM subsystem, it just happens to look for the `wakeup-source` devicetree property. But, it also looks for `gpios`, `interrupts`, etc. and it is not part of those either :) Thus, making it part of pm subsystem would be misleading IMO, 

I think `os/lib` could be the right place for it, like `sys_poweroff()` and `sys_reboot()`. subsys is not the correct place for it I think either, neither is subsys/pm IMO

ceolin commented:
> > @bjarki-trackunit I do not think this should be submitted as a new subsystem, please fold this under subsys/pm somehow, this is all about power management after all, it does not need to depend on current PM infrastrucutre.
> 
> The wakeup source subsystem does not manage devices like the pm subsystem, it configures the SOC to wake up on events. It is an extension of the SOC capabilities, not an addition to device drivers. Its similar to `sys_poweroff()` which configures the SOC to go to sleep.

The `pm` subsystem is not only about device drivers, it is also about SoC pm.
> 
> Like `sys_poweroff()`, it does not depend on, or is even related to the PM subsystem, it just happens to look for the `wakeup-source` devicetree property. But, it also looks for `gpios`, `interrupts`, etc. and it is not part of those either :) Thus, making it part of pm subsystem would be misleading IMO.

I kind of disagree, I think that is totally related to PM. poweroff / reboot one can say that is the application who will use them and there are targets without PM ....  While this latter part is true here, wakeup source is target specific. Also, the pm subsystem is already using this property to not suspend devices that are used as wake up source.

The pm subsytem could use it to configure sources when soc is suspending or something like that.
That's said, we don't need to bring PM infra when this is enabled, same way we have independent SoC / device PM.

> 
> I think `lib/os` could be the right place for it, like `sys_poweroff()` and `sys_reboot()`. `subsys` is not the correct place for it either I think, neither is `subsys/pm`.

bjarki-trackunit commented:
@ceolin I updated the description with a glossary to make it easier to discuss :)

A wakeup source is not tied to a struct device, it is tied to a devicetree node (See example in overview with the `user_button` node)

Enabling a wakeup source means enabling all wakeup events said wakeup source can generate, which happens right before poweroff when the system is in a safe state. The flow is as follows:

1. Application enables wakeup sources using `sys_wakeup_source_enable()` (which just sets a flag)
2. Application suspends devices using PM subsystem (if enabled)
3. Application calls `sys_poweroff()`
4. `sys_poweroff()` triggers wakeup source subsystem to configure all wakeup events defined by enabled wakeup sources
5. Wakeup event (or reset) wakes up SOC

The wakeup subsystem only depends on:
* `sys_poweroff()`
* SoC implementations of the `z_wakeup_source_ ...` hooks
* Devicetree identifers of wakeup events (named IRQs, gpios, vendor specific id)

The application is free to resume and suspend any device, independent of if said devices shall be able to wake up the system. You may want to keep a GNSS device in a low power mode to maintain FIX while not wanting it to wake up the SoC for example.

bjarki-trackunit commented:
@henrikbrixandersen This is how Linux solves the interrupt GPIO conundrom:

Linux example:
```
	&spi0 {
		sensor: bmp085@0 {
			compatible = "bosch,bmp085"
			reg = <0>;
			interrupt-parent = <&gpio0>;
			interrupts = <11 IRQ_TYPE_EDGE_RISING>;
			reset-gpios = <&gpio0 12 (GPIO_ACTIVE_HIGH)>;
		}
	};
```

Zephyr example:
```
	&spi0 {
		sensor: bmp085@0 {
			compatible = "bosch,bmp085"
			reg = <0>;
			int-gpios = <&gpio0 11 (GPIO_ACTIVE_HIGH)>;
			reset-gpios = <&gpio0 12 (GPIO_ACTIVE_HIGH)>;
		}
	};
```

In Linux, the GPIO controller can be used as an interrupt controller, in which case it becomes quite obvious which GPIO pin to enable as wakeup source, we can even use the named approach:
```
	&spi0 {
		sensor: bmp085@0 {
			compatible = "bosch,bmp085"
			reg = <0>;
			interrupt-parent = <&gpio0>;
			interrupts = <11 IRQ_TYPE_EDGE_RISING>;
			interrupt-names = "wakeup";
			reset-gpios = <&gpio0 12 (GPIO_ACTIVE_HIGH)>;
		}
	};
```

Do you think we should go the same route in Zephyr? given that we are already using the "pattern" `int-gpios` or similar to indicate that it is a routed interrupt?

bjarki-andreasen committed:
scripts: dts: edtlib: Allow phandle names for *-gpios

There is a special casing for *-gpios properties which maps all
these properties to the specifer space "gpio". This special
casing however is also applied to the phandle names currently,
which makes all *-gpios properties like int-gpios map to
gpio-names, instead of int-gpio-names.

This commit reverts the *-gpios special casing before adding
phandle names, allowing int-gpio-names to map to int-gpios.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
dts: bindings: base: Add interrupt-gpios base property

Add interrupt-gpios property which will be used to specify
interrupts routed through GPIOs.

This is useful for the wakeup source subsystem which needs to
identify these gpios to enable them as wakeup sources.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
devicetree: gpio: Add GPIO_*_BY_NAME macros

The interrupt-gpios are addressable by name, like the
interrupts property, using the new property
interrupt-gpio-names.

Add macros to get elements in gpios phandle-arrays by name.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

henrikbrixandersen reviewed (changes requested):
> Add interrupt-gpios and interrupt-gpio-names

I very strongly believe that this is not the way to go about this.

bjarki-trackunit commented:
> > Add interrupt-gpios and interrupt-gpio-names
> 
> I very strongly believe that this is not the way to go about this.

I will look into using GPIO controllers as interrupt-parents as well, then we have something to compare :)

bjarki-trackunit commented:
This PR is currently blocked by the interrupt rework, as it requires a way to configure gpio routed interrupts, hopefully it will start moving again soon :)
